src/realm/group_writer.cpp

```diff
@@ -498,54 +515,79 @@ size_t GroupWriter::get_free_space() {
+        m_free_positions.add(e.ref);
+        m_free_lengths.add(e.size);
+        if (is_shared)
+            m_free_versions.add(e.released_at_version);
+    }
+    m_free_in_file.clear();
+}
 
-    size_t n = m_free_lengths.size() - 1;
-    for (size_t i = 0; i < n; ++i) {
-        size_t i2 = i + 1;
-        size_t pos1 = to_size_t(m_free_positions.get(i));
-        size_t size1 = to_size_t(m_free_lengths.get(i));
-        size_t pos2 = to_size_t(m_free_positions.get(i2));
-        REALM_ASSERT_RELEASE_EX(pos1 + size1 <= pos2, pos1, size1, pos2);
-        if (pos2 == pos1 + size1) {
-            // If this is a shared db, we can only merge
-            // segments where no part is currently in use
-            if (is_shared) {
-                size_t v1 = to_size_t(m_free_versions.get(i));
-                if (v1 >= m_readlock_version)
-                    continue;
-                size_t v2 = to_size_t(m_free_versions.get(i2));
-                if (v2 >= m_readlock_version)
-                    continue;
-            }
+void GroupWriter::sort_freelist()
+{
+    std::sort(begin(m_free_in_file), end(m_free_in_file), [](auto& a, auto& b) { return a.ref < b.ref; });
+}
 
-            // FIXME: Performing a series of calls to Array::erase() is
-            // unnecessarily expensive because we have to shift the contents
-            // above i2 multiple times in general. A more efficient way would be
-            // to use assigments only, and then make a final call to
-            // Array::truncate() when the new size is know.
-
-            // Merge
-            size_t size2 = to_size_t(m_free_lengths.get(i2));
-            size_t new_size = size1 + size2;
-            REALM_ASSERT_RELEASE_EX(new_size > size1 && new_size > size2, size1, size2, new_size);
-            m_free_lengths.set(i, new_size);
-            m_free_positions.erase(i2);
-            m_free_lengths.erase(i2);
-            if (is_shared)
-                m_free_versions.erase(i2);
-
-            --n;
-            --i; // May underflow, but that is ok
+void GroupWriter::merge_adjacent_entries_in_freelist()
+{
+    if (m_free_in_file.size() == 0)
+        return;
+    // Combine any adjacent chunks in the freelist
+    auto prev = m_free_in_file.begin();
+    for (auto it = m_free_in_file.begin() + 1; it != m_free_in_file.end(); ++it) {
+        if (prev->ref + prev->size != it->ref) {
+            prev = it;
+            continue;
         }
+
+        prev->size += it->size;
+        it->size = 0;
     }
 }
 
+void GroupWriter::filter_empty_entries_in_freelist()
+{
+    // Remove all of the now zero-size chunks from the free list
+    m_free_in_file.erase(
+        std::remove_if(begin(m_free_in_file), end(m_free_in_file), [](auto& chunk) { return chunk.size == 0; }),
+        end(m_free_in_file));
+}
+
+void GroupWriter::merge_free_space()
+{
+    sort_freelist();
+    merge_adjacent_entries_in_freelist();
+    filter_empty_entries_in_freelist();
```

Comment:
No need to filter out empty entries here. That can be done when writing the list back.

Reply:
Yeah, correct, but in a later iteration I'll want to only use the std::vector for allocation as well.